controllers/clusters/on_premises.go

```diff
@@ -93,102 +90,154 @@ func handleCreateOnPremisesClusterResources(ctx context.Context, b *onPremisesBo
 	return nil
 }
 
+func handleNodeReplace(ctx context.Context, b *onPremisesBootstrap, newNode, oldNode *v1beta1.Node) error {
+	if len(b.ClusterStatus.DataCentres) == 0 {
+		return models.ErrZeroDataCentres
+	}
+
+	objectName := b.K8sObject.GetName()
+	vm := &virtcorev1.VirtualMachine{}
+	for i := 0; i < len(b.ClusterStatus.DataCentres[0].Nodes); i++ {
+		nodeName := fmt.Sprintf("%s-%d-%s", models.NodeVMPrefix, i, objectName)
+		err := b.K8sClient.Get(ctx, types.NamespacedName{
+			Namespace: b.K8sObject.GetNamespace(),
+			Name:      nodeName,
+		}, vm)
+		if client.IgnoreNotFound(err) != nil {
+			return err
+		}
+		if err == nil {
+			continue
+		}
+
+		osDVName := fmt.Sprintf("%s-%d-%s", models.NodeOSDVPrefix, i, objectName)
+		osDV, err := reconcileDV(ctx, b, newNode.ID, osDVName, models.OSDVLabel)
+		if err != nil {
+			return err
+		}
+
+		storageDVList := &cdiv1beta1.DataVolumeList{}
+		err = b.K8sClient.List(ctx, storageDVList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				models.NodeIDLabel: oldNode.ID,
+				models.DVRoleLabel: models.StorageDVLabel,
+			}),
+		})
+		if err != nil {
+			return err
+		}
+
+		storageDV := &cdiv1beta1.DataVolume{}
+		storageDVName := fmt.Sprintf("%s-%d-%s", models.NodeDVPrefix, i, objectName)
+		err = b.K8sClient.Get(ctx, types.NamespacedName{
+			Namespace: b.K8sObject.GetNamespace(),
+			Name:      storageDVName,
+		}, storageDV)
+		if err != nil {
+			return err
+		}
+
+		patch := client.MergeFrom(storageDV.DeepCopy())
+		storageDV.Labels[models.NodeIDLabel] = newNode.ID
+		err = b.K8sClient.Patch(ctx, storageDV, patch)
+		if err != nil {
+			return err
+		}
+
+		svc := &k8scorev1.Service{}
+		svcName := fmt.Sprintf("%s-%s", models.ClusterIPServiceRoleLabel, nodeName)
+		err = b.K8sClient.Get(ctx, types.NamespacedName{
+			Namespace: b.K8sObject.GetNamespace(),
+			Name:      svcName,
+		}, svc)
+		if err != nil {
+			return err
+		}
+
+		patch = client.MergeFrom(svc.DeepCopy())
+		svc.Labels[models.NodeIDLabel] = newNode.ID
+		err = b.K8sClient.Patch(ctx, svc, patch)
+		if err != nil {
+			return err
+		}
+
+		_, err = reconcileVM(ctx, b, newNode.ID, nodeName, newNode.Rack, osDV.Name, storageDV.Name)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	return models.ErrGenerateVMName
+}
```

Comment:
If there is no node to replace, we return an error - I think this is a no good way to end the func? At least return an error like "No node to replace".

Reply:
Changed code a bit, now should look better